packages/Meshfree/src/DTK_SplineOperator_def.hpp

```diff
@@ -0,0 +1,300 @@
+Teuchos::RCP<
+    typename SplineOperator<DeviceType, CompactlySupportedRadialBasisFunction,
+                            PolynomialBasis>::Operator>
+SplineOperator<DeviceType, CompactlySupportedRadialBasisFunction,
+               PolynomialBasis>::
+    buildBasisOperator(
+        Teuchos::RCP<const Map> domain_map, Teuchos::RCP<const Map> range_map,
+        Kokkos::View<Coordinate const **, DeviceType> source_points,
+        Kokkos::View<Coordinate const **, DeviceType> target_points,
+        int const knn )
+{
+    auto teuchos_comm = domain_map->getComm();
+    auto teuchos_mpi_comm =
+        Teuchos::rcp_dynamic_cast<const Teuchos::MpiComm<int>>( teuchos_comm );
+    MPI_Comm comm = ( *teuchos_mpi_comm->getRawMpiComm() )();
+
+    int const num_source_points = source_points.extent( 0 );
+    int const num_points = target_points.extent( 0 );
+
+    ArborX::DistributedSearchTree<DeviceType> distributed_tree( comm,
+                                                                source_points );
+    DTK_CHECK( !distributed_tree.empty() );
+
+    // Perform the actual search.
+    auto queries =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::makeKNNQueries(
+            target_points, knn );
+
+    Kokkos::View<int *, DeviceType> offset( "offset", 0 );
+    Kokkos::View<int *, DeviceType> ranks( "ranks", 0 );
+    Kokkos::View<int *, DeviceType> indices( "indices", 0 );
+    distributed_tree.query( queries, indices, offset, ranks );
+
+    // Retrieve the coordinates of all points that met the predicates.
+    auto source_points_with_halo =
+        Details::NearestNeighborOperatorImpl<DeviceType>::fetch(
+            comm, ranks, indices, source_points );
+
+    auto transformed_source_points = Details::MovingLeastSquaresOperatorImpl<
+        DeviceType>::transformSourceCoordinates( source_points_with_halo,
+                                                 offset, target_points );
+
+    // To build the radial basis function, we need to define the radius of
+    // the radial basis function. Since we use kNN, we need to compute the
+    // radius. We only need the coordinates of the source points because of
+    // the transformation of the coordinates.
+    auto radius =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::computeRadius(
+            transformed_source_points, offset );
+
+    // Build phi (weight matrix)
+    auto phi =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::computeWeights(
+            transformed_source_points, radius,
+            CompactlySupportedRadialBasisFunction() );
+
+    // Compute some helper arrays
+    int rank_offset = 0;
+    MPI_Scan( &num_source_points, &rank_offset, 1, MPI_INT, MPI_SUM, comm );
+
+    int comm_size = teuchos_comm->getSize();
+
+    std::vector<int> cumulative_points_per_process( comm_size + 1 );
+    MPI_Allgather( &rank_offset, 1, MPI_INT,
+                   &( cumulative_points_per_process[1] ), 1, MPI_INT, comm );
```

Comment:
Not sure I get this.  Do you need to do a exclusive scan?

Reply:
Why do we just do a all gather and then locally a scan?

Reply:
This sounds like a good idea.